src/controlador.ts

```diff
@@ -1,49 +1,93 @@
-    // Lógica para obtener todas las tareas
+  constructor(private readonly optimizador: OptimizadorSemanal) {}
 
-    this.logger.debug('Obteniendo todas las tareas programadas');
-    return tareas;
+  @Get()
+  obtenerTodasLasTareas(): Actividad[] | undefined {
+    try {
+      const tareas: Actividad[] = this.optimizador.Actividades;
+      this.logger.debug('Obteniendo todas las tareas programadas');
+      return tareas;
+    } catch (error) {
+      this.handleException(error);
+      return undefined;
+    }
   }
 
   @Post()
-  crearTarea(): string {
-    // Lógica para crear una nueva tarea
+  @UseFilters(new ErrorFilter())
+  crearTarea(@Body() body: any): Actividad | undefined {
+    try {
+      const { descripcion, duracion } = body;
 
-    this.logger.info('Tarea creada');
-    return 'Tarea creada';
+      if (descripcion == null || duracion == null) {
+        throw new HttpException('La descripción y la duración son obligatorias para crear una tarea', HttpStatus.BAD_REQUEST);
+      }
+
+      const actividadVariable = new ActividadVariable(descripcion, this.logger, duracion);
+      this.logger.info('Tarea creada');
+      return actividadVariable;
+    } catch (error) {
+      this.handleException(error);
+      return undefined;
+    }
   }
 
   @Get(':id')
-  obtenerTareaPorId(@Param('id') id: string): string {
-    // Lógica para obtener una tarea por su ID
+  obtenerTareaPorId(@Param('id') id: ActividadId): Actividad | undefined {
+    try {
+      const actividad = this.optimizador.Actividades.find((act) => act.Id === id);
+
+      if (!actividad) {
+        throw new HttpException(`No se encontró la tarea con id: ${id}`, HttpStatus.NOT_FOUND);
+      }
 
-    this.logger.debug(`Obteniendo tarea con ID ${id}`);
-    return `Tarea con ID ${id}`;
+      this.logger.info(`Tarea con id ${id} encontrada`);
+      return actividad;
+    } catch (error) {
+      this.handleException(error);
+      return undefined;
+    }
   }
 
   @Put(':id')
-  actualizarTarea(@Param('id') id: string): string {
-    // Lógica para actualizar una tarea por su ID
-
-    this.logger.info(`Tarea con ID ${id} actualizada`);
-    return `Tarea con ID ${id} actualizada`;
+  @UseFilters(new ErrorFilter())
+  actualizarTarea(@Param('id') id: ActividadId): string | undefined {
+    try {
+      // Lógica para actualizar una tarea por su ID
+      this.logger.info(`Tarea con ID ${id} actualizada`);
+      return `Tarea con ID ${id} actualizada`;
+    } catch (error) {
+      this.handleException(error);
+      return undefined;
+    }
   }
 
-  @Delete(':id')
-  eliminarTarea(@Param('id') id: string): string {
-    // Lógica para eliminar una tarea por su ID
-
-    this.logger.info(`Tarea con ID ${id} eliminada`);
-    return `Tarea con ID ${id} eliminada`;
+  private handleException(error: any): void {
+    this.logger.error(error.message);
+    throw new HttpException('Error interno del servidor', HttpStatus.INTERNAL_SERVER_ERROR);
```

Comment:
Por favor, mira [los errores frecuentes de este objetivo](https://github.com/JJ/IV-/blob/master/errores/errores-objetivo-9.md):

> Devolver un error 500 cuando suceda cualquier tipo de problema sin una clasificación clara. Una aplicación nunca puede devolver error 500, porque este indica que la aplicación no funciona correctamente y está en un estado en el que no puede funcionar. Tratad de buscar un error que sea más adecuado, que puede ser desde 400 Bad Request (que indica que algo va mal, pero no se sabe específicamente qué) o 422 Unprocessable content o, en general, algo que sea más apropiado para la excepción que se haya podido generar.

Reply:
Los errores específicos como la falta de descripción, o el no encontrar id se manejan específicamente:
```
      if (descripcion == null || duracion == null) {
        throw new HttpException('La descripción y la duración son obligatorias para crear una tarea', HttpStatus.BAD_REQUEST);
      }
```

```
      if (!actividad) {
        throw new HttpException(`No se encontró la tarea con id: ${id}`, HttpStatus.NOT_FOUND);
      }
```

El error 500 solo debería mostrarse si no es uno de esos errores, es decir, ha habido un problema al ejecutar la solicitud pero por problemas internos. En realidad, no sé qué problemas son, entonces no sé como hacerlos más específicos.

Reply:
Una aplicación no debería devolver *nunca* un error 500, porque este error indica simplemente que la aplicación no funciona y que no debe hacerse ninguna interacción adicional con ella, porque está en un estado incorrecto... O ha dejado de funcionar, porque es el error que suele devolverse desde un proxy o algún servidor *delante* de tu aplicación. En los status 4xx tienes errores genéricos como [*bad request*](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400) que viene a decir "Oye, hay un problema, pero no sé muy bien de qué se trata, si lo supiera te habría devuelto un 404 o un 418"). 
Hay errores 500 específicos que sí se pueden devolver de una aplicación, como el [501](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501)  (no implementado). En general, los 4xy son "hay un error, parece culpa tuya". Los 5xy son "hay un error, es culpa mía".  No parece que sea el caso aquí. Es más bien "Hay un error, pero no logro encuadrarlo en los que tengo ya clasificados, pero en todo caso es culpa tuya, yo no he hecho hada". O sea, un 4xx

Reply:
Genial, lo cambio. Muchas gracias por la explicación.

Reply:
Un poco de [info adicional aquí](https://stackoverflow.com/questions/57709424/is-it-good-to-return-http-500-in-a-restful-api). Yo creo que la clave está en esta frase:

> Certainly, you shouldn't be using a 500 status code when the fault is in the request. Distinguishing 4xx from 5xx is a lot more important than distinguishing 500 from 50x.

Creo que es el caso que se aplica aquí. Si es un error en tu lógica de negocio, por ejemplo, se podría devolver (aunque en la respuesta dicen que sería más razonable un error temporal, quizás 503), pero si es un error en una petición que no ha sido clasificado de otra forma, parece más razonable usar un 400.

Reply:
> Genial, lo cambio. Muchas gracias por la explicación.

A ti por preguntar por qué. Ya sabes que es la clave para ser una buena ingeniera :-)